great_expectations/datasource/fluent/interfaces.py

```diff
@@ -257,7 +257,10 @@ def _validate_batch_request(self, batch_request: BatchRequest) -> None:
 
     @public_api
     def add_batch_definition(
-        self, name: str, partitioner: Optional[Partitioner] = None
+        self,
+        name: str,
+        partitioner: Optional[Partitioner] = None,
+        batching_regex: Optional[re.Pattern] = None,
```

Comment:
This seems too high in the hierarchy since SQL DataAssets will also inherit from here and don't use the `batching_regex`.

Do we need a generic `add_batch_definition` now that we have specific ones like `add_batch_definition_daily`? Or maybe we remove the public_decorator api if we want some general functionality here. It would still feels odd to have the `batching_regex` argument since that doesn't apply to all assets. I know we only have 1 BatchDefinition class now so it might be ok if this wasn't public but that punts the work to putting it where it should be if we introduce multiple of these.

Reply:
yeah, that's a good point - i think the best path forward is to use this as the canonical way to add a batch def, but remove it from the public api

Reply:
Definitely agree on removing `public_api`. But I agree with @billdirks  that the best long term solution is not to have any reference to batch regex at this layer.

My understanding from earlier this week was that just like day/month/year + column name defines how to partition sql, batching_regex (with groups for day/month/year) defines how to partition file-based, which I thought meant that batching_regex gets wrapped in pandas/spark partitioner. Maybe this is an implementation detail I'm off on though.

Reply:
thanks @tyler-hoffman! `batching_regex` doesn't get wrapped into a partitioner, although it has a similar function. If we have Asset-specific BatchDefinition objects -- say `SqlBatchDefinition` -- then it makes sense to just have a protocol for `add_batch_definition` top level and, and each branch of the inheritance chain needs to implement it. But since we have a single type of BatchDefinition it feels a bit strange to me to exclude a property from the signature, even if some inheriting classes won't use it. 

To put it another way, the responsibility of this function is to construct a BatchDefinition, which has a `batching_regex` property, so it makes sense to me that this base method would have it as well.

Reply:
> doesn't get wrapped into a partitioner
Is there a reason not to do this, though? We've already branched on on partitioners, so wouldn't it make sense to put them there?

My concerns are:
* A nullable field that only gets used by some implementations feels like a (slight) smell. It's like the parent class knowing about specific child classes, which shouldn't be the case. Why would the base class even know about batching regex? Similar to why would they know about column_name that only gets used for SQL
* A nullable field requires more runtime checking, and the types don't tell engineers what to expect
* It seems more clear to colocate the mechanism for determining year/month/day/etc with type of sorting. i.e. then it parallels column_name + partitioner type